src/common/meta/src/kv_backend/chroot.rs

```diff
@@ -136,121 +136,138 @@ impl KvBackend for ChrootKvBackend {
-        let mut new_end = root.to_vec();
-        let mut ok = false;
-        for i in (0..new_end.len()).rev() {
-            new_end[i] = new_end[i].wrapping_add(1);
-            if new_end[i] != 0 {
-                ok = true;
-                break;
+    // see namespace.prefixInterval - https://github.com/etcd-io/etcd/blob/v3.5.10/client/v3/namespace/util.go
+    fn range_end_prepend_root(&self, mut range_end: Vec<u8>) -> Vec<u8> {
+        let root = &self.root;
+        if range_end == [0] {
+            // the edge of the keyspace
+            let mut new_end = root.clone();
+            let mut ok = false;
+            for i in (0..new_end.len()).rev() {
+                new_end[i] = new_end[i].wrapping_add(1);
+                if new_end[i] != 0 {
+                    ok = true;
+                    break;
+                }
             }
+            if !ok {
+                // 0xff..ff => 0x00
+                new_end = vec![0];
+            }
+            new_end
+        } else if !range_end.is_empty() {
+            let mut new_end = root.clone();
+            new_end.append(&mut range_end);
+            new_end
+        } else {
+            vec![]
         }
-        if !ok {
-            // 0xff..ff => 0x00
-            new_end = vec![0];
-        }
-        new_end
-    } else if !range_end.is_empty() {
-        let mut new_end = root.to_vec();
-        new_end.append(&mut range_end);
-        new_end
-    } else {
-        vec![]
+    }
+
+    fn txn_prepend_root(&self, mut txn: Txn) -> Txn {
+        let op_prepend_root = |op: TxnOp| match op {
+            TxnOp::Put(k, v) => TxnOp::Put(self.key_prepend_root(k), v),
+            TxnOp::Get(k) => TxnOp::Get(self.key_prepend_root(k)),
+            TxnOp::Delete(k) => TxnOp::Delete(self.key_prepend_root(k)),
+        };
+        txn.req.success = txn.req.success.drain(..).map(op_prepend_root).collect();
+        txn.req.failure = txn.req.failure.drain(..).map(op_prepend_root).collect();
+        txn.req.compare = txn
+            .req
+            .compare
+            .drain(..)
+            .map(|cmp| super::txn::Compare {
+                key: self.key_prepend_root(cmp.key),
+                cmp: cmp.cmp,
+                target: cmp.target,
+            })
+            .collect();
+        txn
     }
 }
 
-fn txn_prepend_root(root: &[u8], mut txn: Txn) -> Txn {
-    fn op_prepend_root(root: &[u8], op: TxnOp) -> TxnOp {
-        match op {
-            TxnOp::Put(k, v) => TxnOp::Put(key_prepend_root(root, k), v),
-            TxnOp::Get(k) => TxnOp::Get(key_prepend_root(root, k)),
-            TxnOp::Delete(k) => TxnOp::Delete(key_prepend_root(root, k)),
+#[cfg(test)]
+mod tests {
+    use std::sync::Arc;
+
+    use crate::kv_backend::chroot::ChrootKvBackend;
+    use crate::kv_backend::memory::MemoryKvBackend;
+
+    #[test]
+    fn test_prefix_key_and_range_end() {
+        fn run_test_case(pfx: &[u8], key: &[u8], end: &[u8], w_key: &[u8], w_end: &[u8]) {
```

Comment:
Nit: I prefer `prefix` instead of `pfx`. And what's the meaning of the `w_end`? Any other good name?

Reply:
Yep. This is directly copy from Golang's impl. `w_end` actually means `expected_end`. I don't change the flavor in the first place.

If it's good to do, we can apply it. But could you pull my PR can push to update it? I'm travelling this day so I cannot push new commit perhaps. Just do what you want.

Reply:
It's okay; it's just some non-blocking nit.